lib/plugin.js

```diff
@@ -0,0 +1,102 @@
+'use strict'
+
+const debug = require('debug')('token:renew:plugin')
+const Try = require('flat-try')
+
+const THRESHOLD = 60 * 60 * 1000 // renew one hour before expiring
+
+class TokenRenewalPlugin {
+  /**
+   * @param {Object} adapter
+   * @param {function: Promise<{ authToken: string, expiresAt: number }>} adapter.refreshToken
+   */
+  constructor (adapter) {
+    this.id = 'renew-token-plugin'
+    this.type = 'ws2'
+    this.manager = {
+      'ws:created': this._onCreate.bind(this),
+      'ws:destroyed': this._onClose.bind(this)
+    }
+    this._algoManagers = new Map()
+    this.adapter = adapter
+
+    this._renewAuthToken = this._renewAuthToken.bind(this)
+  }
+
+  close () {
+    this._clearTimeout()
+    this._algoManagers.clear()
+  }
+
+  _onCreate (args = {}) {
+    const { id, manager, state = {} } = args
+    this._algoManagers.set(id, new WeakRef(manager))
+
+    if (!this._timeout) {
+      this._scheduleAutoRenewal()
+    }
+
+    return state
+  }
+
+  _onClose ({ id, state = {} } = {}) {
+    this._algoManagers.delete(id)
+
+    if (this._algoManagers.size === 0) {
+      this._clearTimeout()
+    }
+
+    return state
+  }
+
+  _scheduleAutoRenewal (expiresAt = 0) {
+    const timeout = expiresAt - Date.now() - THRESHOLD
+    this._timeout = setTimeout(this._renewAuthToken, timeout)
+  }
+
+  async _renewAuthToken () {
+    this._clearTimeout()
+
+    const [err, response] = await Try.promise(() => this.adapter.refreshToken())
+
+    if (err) {
+      debug('failed to renew auth token: %j', err)
+
+      for (const manager of this._getManagers()) {
+        manager.emit('plugin:error', `[${this.id}] error: ${err.message}`)
+      }
+
+      this._scheduleAutoRenewal()
+      return
+    }
+
+    const { authToken, expiresAt } = response
+
+    for (const manager of this._getManagers()) {
+      manager.auth({ authToken })
+    }
+
+    this._scheduleAutoRenewal(expiresAt)
+  }
+
+  * _getManagers () {
```

Comment:
is it possible to get rid of the generator to simplify the code?

Reply:
30ba35e73fc51594de07e3ddc2b4ab64ffbd18a8